### ISSUE TITLE
ZipArchiver::supports(): Argument #2 ($sourceType) must be of type string, null given

### ISSUE THREAD

**Issue by mikkoaf** [composer/composer#10702]
Output of `composer diagnose`:

```
$ composer diagnose
Checking composer.json: WARNING
The version field is present, it is recommended to leave it out if the package is published on Packagist.
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys:
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 2.3.3
PHP version: 8.0.17
PHP binary path: /usr/bin/php8.0
OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
cURL version: 7.68.0 libz 1.2.11 ssl OpenSSL/1.1.1f
zip: extension present, unzip present, 7-Zip not available
```

When I run this command: 

```
composer archive --format=zip --file=composer-archive
```

I get the following output: 

```
$ composer archive --format=zip --file=composer-archive
Creating the archive into ".".

In ZipArchiver.php line 81:

  [TypeError]
  Composer\Package\Archiver\ZipArchiver::supports(): Argument #2 ($sourceType) must be of type string, null given, called in phar:///usr/local/bin/composer/src/Composer/Package/Archiver/ArchiveManager.php o
  n line 133


Exception trace:
  at phar:///usr/local/bin/composer/src/Composer/Package/Archiver/ZipArchiver.php:81
 Composer\Package\Archiver\ZipArchiver->supports() at phar:///usr/local/bin/composer/src/Composer/Package/Archiver/ArchiveManager.php:133
 Composer\Package\Archiver\ArchiveManager->archive() at phar:///usr/local/bin/composer/src/Composer/Command/ArchiveCommand.php:138
 Composer\Command\ArchiveCommand->archive() at phar:///usr/local/bin/composer/src/Composer/Command/ArchiveCommand.php:94
 Composer\Command\ArchiveCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:298
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:1015
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:299
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:334
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:130
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:83
 require() at /usr/local/bin/composer:29

archive [-f|--format FORMAT] [--dir DIR] [--file FILE] [--ignore-filters] [--] [<package> [<version>]]
```

**Comment by tubbynl** [comment]
this occurred in my CI-build where i used `composer:2` as build image. 

* error also occurred on version 2.3.3
* error also occurred on version 2.3.4
* error doesnt occur on version 2.2.11

i'm unsure what composer image version was used in on the original :2 version spec but perhaps this helps

my stacktrace

```
$ composer archive
Installing dependencies from lock file
Verifying lock file contents can be installed on current platform.
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update <package name>`.
Package operations: 0 installs, 0 updates, 33 removals
  - Removing webmozart/assert (1.10.0)
...
  - Removing doctrine/instantiator (1.4.1)
Generating optimized autoload files
30 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
Creating the archive into ".".
In ZipArchiver.php line 81:
                                                                               
  [TypeError]                                                                  
  Composer\Package\Archiver\ZipArchiver::supports(): Argument #2 ($sourceType  
  ) must be of type string, null given, called in phar:///usr/bin/composer/sr  
  c/Composer/Package/Archiver/ArchiveManager.php on line 133                   
                                                                               
Exception trace:
  at phar:///usr/bin/composer/src/Composer/Package/Archiver/ZipArchiver.php:81
 Composer\Package\Archiver\ZipArchiver->supports() at phar:///usr/bin/composer/src/Composer/Package/Archiver/ArchiveManager.php:133
 Composer\Package\Archiver\ArchiveManager->archive() at phar:///usr/bin/composer/src/Composer/Command/ArchiveCommand.php:138
 Composer\Command\ArchiveCommand->archive() at phar:///usr/bin/composer/src/Composer/Command/ArchiveCommand.php:94
 Composer\Command\ArchiveCommand->execute() at phar:///usr/bin/composer/vendor/symfony/console/Command/Command.php:298
 Symfony\Component\Console\Command\Command->run() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:1015
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:299
 Symfony\Component\Console\Application->doRun() at phar:///usr/bin/composer/src/Composer/Console/Application.php:334
 Composer\Console\Application->doRun() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at phar:///usr/bin/composer/src/Composer/Console/Application.php:130
 Composer\Console\Application->run() at phar:///usr/bin/composer/bin/composer:83
 require() at /usr/bin/composer:29
archive [-f|--format FORMAT] [--dir DIR] [--file FILE] [--ignore-filters] [--] [<package> [<version>]]
```

**Comment by tubbynl** [comment]
ok; locally also reproducable when no `version` spec in the composer.json composer archive works; if i spec a version then it fails with the same stacktrace (using composer 2.3.3). Let me fix a example composer json file

**Comment by tubbynl** [comment]
composer.json
```json
{
    "name": "coloss/concept500",
    "support": {
        "email": "github@tubby.nl"
    },
    "require": {
        "php": ">=8.1"
    },
    "require-dev": {
        "phpunit/phpunit": "^9"
    },
    "config": {
        "archive-format": "zip"
    },
    "archive": {
        "name": "test"
    }
}
```

```bash
composer archive
# fails
git init
git add .
git commit -m "init"
composer archive
# works
composer config version 1.0.0
# fails
```

**Comment by tubbynl** [comment]
i can verify it is fixed in 2.3.4 , tnx @Seldaek !
